File: xaiop-sdk/python/src/xaiop/stream/transport.py
"""Network transports for XaiopStream."""

from __future__ import annotations

import codecs
import http.client
import threading
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from typing import Any, Callable, Iterable, Iterator

TRANSPORT_KIND = {
    "HTTP": "http",
    "SSE": "sse",
    "RAW": "raw",
}

TransportKind = str


@dataclass
class TransportRequest:
    url: str = ""
    transport: TransportKind = TRANSPORT_KIND["HTTP"]
    method: str = "GET"
    headers: dict[str, str] | None = None
    body: str | bytes | None = None
    timeout_ms: int | None = None
    sse_events: set[str] | None = None
    source: Iterable[str | bytes] | None = None
    aborted: threading.Event | None = None


@dataclass
class TransportHandlers:
    on_text: Callable[[str], None]
    on_done: Callable[[], None]
    on_error: Callable[[Exception], None]


class TransportHandle:
    def __init__(self, abort_fn: Callable[[], None]) -> None:
        self.abort = abort_fn


def open_transport(
    req: TransportRequest | dict[str, Any],
    handlers: TransportHandlers | dict[str, Any],
) -> TransportHandle:
    if isinstance(req, dict):
        req = TransportRequest(**_normalize_transport_req(req))
    if isinstance(handlers, dict):
        handlers = TransportHandlers(
            on_text=handlers["onText"] if "onText" in handlers else handlers["on_text"],
            on_done=handlers["onDone"] if "onDone" in handlers else handlers["on_done"],
            on_error=handlers["onError"] if "onError" in handlers else handlers["on_error"],
        )

    aborted = req.aborted or threading.Event()

    def _abort() -> None:
        aborted.set()

    def _run() -> None:
        try:
            kind = req.transport or TRANSPORT_KIND["HTTP"]
            if kind == TRANSPORT_KIND["RAW"]:
                _run_raw(req, handlers, aborted)
            elif kind == TRANSPORT_KIND["SSE"]:
                _run_sse(req, handlers, aborted)
            else:
                _run_http(req, handlers, aborted)
            if not aborted.is_set():
                handlers.on_done()
        except Exception as err:
            if aborted.is_set():
                handlers.on_error(Exception("aborted"))
            else:
                handlers.on_error(err if isinstance(err, Exception) else Exception(str(err)))

    thread = threading.Thread(target=_run, name="xaiop-transport", daemon=True)
    thread.start()
    return TransportHandle(_abort)


def _normalize_transport_req(req: dict[str, Any]) -> dict[str, Any]:
    out = dict(req)
    if "transport" not in out and "kind" in out:
        out["transport"] = out.pop("kind")
    return out


def _emit_text(handlers: TransportHandlers, text: str) -> None:
    if text:
        handlers.on_text(text)


def _run_raw(
    req: TransportRequest,
    handlers: TransportHandlers,
    aborted: threading.Event,
) -> None:
    if req.source is None:
        raise TypeError("raw transport requires `source`")
    decoder = codecs.getincrementaldecoder("utf-8")()
    for piece in req.source:
        if aborted.is_set():
            raise Exception("aborted")
        if piece is None:
            continue
        if isinstance(piece, str):
            _emit_text(handlers, piece)
        elif isinstance(piece, (bytes, bytearray)):
            _emit_text(handlers, decoder.decode(bytes(piece)))
        else:
            raise TypeError("raw source yielded unsupported chunk type")
    tail = decoder.decode(b"", final=True)
    _emit_text(handlers, tail)


def _run_http(
    req: TransportRequest,
    handlers: TransportHandlers,
    aborted: threading.Event,
) -> None:
    try:
        import httpx  # type: ignore
    except ImportError:
        httpx = None  # type: ignore

    if httpx is not None:
        _run_http_httpx(req, handlers, aborted, httpx)
        return
    _run_http_stdlib(req, handlers, aborted)


def _run_http_httpx(req, handlers, aborted, httpx) -> None:
    timeout = None
    if req.timeout_ms and req.timeout_ms > 0:
        timeout = req.timeout_ms / 1000.0
    headers = dict(req.headers or {})
    with httpx.Client(timeout=timeout) as client:
        with client.stream(
            req.method or "GET",
            req.url,
            headers=headers,
            content=req.body,
        ) as response:
            if response.status_code < 200 or response.status_code >= 300:
                raise Exception(f"HTTP {response.status_code}")
            for chunk in response.iter_text():
                if aborted.is_set():
                    raise Exception("aborted")
                _emit_text(handlers, chunk)


def _run_http_stdlib(
    req: TransportRequest,
    handlers: TransportHandlers,
    aborted: threading.Event,
) -> None:
    headers = dict(req.headers or {})
    data = req.body
    if isinstance(data, str):
        data = data.encode("utf-8")
    request = urllib.request.Request(
        req.url,
        data=data,
        headers=headers,
        method=(req.method or "GET").upper(),
    )
    try:
        with urllib.request.urlopen(
            request,
            timeout=(req.timeout_ms / 1000.0) if req.timeout_ms else None,
        ) as response:
            decoder = codecs.getincrementaldecoder("utf-8")()
            while True:
                if aborted.is_set():
                    raise Exception("aborted")
                chunk = response.read(8192)
                if not chunk:
                    break
                _emit_text(handlers, decoder.decode(chunk))
            _emit_text(handlers, decoder.decode(b"", final=True))
    except urllib.error.HTTPError as err:
        raise Exception(f"HTTP {err.code} {err.reason}") from err


def _run_sse(
    req: TransportRequest,
    handlers: TransportHandlers,
    aborted: threading.Event,
) -> None:
    headers = {"Accept": "text/event-stream", **(req.headers or {})}
    sse_req = TransportRequest(
        url=req.url,
        transport=TRANSPORT_KIND["HTTP"],
        method=req.method,
        headers=headers,
        body=req.body,
        timeout_ms=req.timeout_ms,
    )
    allow = req.sse_events
    buf = ""

    def on_chunk(text: str) -> None:
        nonlocal buf
        buf += text
        parts = _split_sse_blocks(buf)
        buf = parts.pop() if parts else ""
        for block in parts:
            data = _parse_sse_block(block, allow)
            _emit_sse_data(handlers, data)

    wrapped = TransportHandlers(on_text=on_chunk, on_done=lambda: None, on_error=handlers.on_error)
    _run_http(sse_req, wrapped, aborted)
    if buf.strip():
        data = _parse_sse_block(buf, allow)
        _emit_sse_data(handlers, data)


def _emit_sse_data(handlers: TransportHandlers, data: str) -> None:
    """Emit SSE payload as wire text; ensure a trailing LF so '.' does not glue to next '>'."""
    if not data:
        return
    if not data.endswith("\n"):
        data = data + "\n"
    _emit_text(handlers, data)


def _split_sse_blocks(buf: str) -> list[str]:
    import re

    parts = re.split(r"\r?\n\r?\n", buf)
    return parts


def _parse_sse_block(block: str, allow: set[str] | None) -> str:
    event = "message"
    data_lines: list[str] = []
    for raw in block.splitlines():
        if not raw or raw.startswith(":"):
            continue
        if ":" in raw:
            field, value = raw.split(":", 1)
            if value.startswith(" "):
                value = value[1:]
        else:
            field, value = raw, ""
        if field == "event":
            event = value
        elif field == "data":
            data_lines.append(value)
    if allow is not None and event not in allow:
        return ""
    if not data_lines:
        return ""
    return "\n".join(data_lines)


def chunks_of(*pieces: str) -> Iterator[str]:
    yield from pieces

File: xaiop-sdk/go/xaiop/stream/transport.go
package stream

import (
	"bufio"
	"bytes"
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"sync/atomic"
	"unicode/utf8"
)

// TransportKind selects the ingest transport.
type TransportKind string

const (
	TransportHTTP TransportKind = "http"
	TransportSSE  TransportKind = "sse"
	TransportRAW  TransportKind = "raw"
)

// TransportRequest configures OpenTransport.
type TransportRequest struct {
	URL        string
	Kind       TransportKind
	Method     string
	Headers    map[string]string
	Body       string
	TimeoutMs  int64
	SSEEvents  map[string]struct{} // nil = all events
	Source     any                 // RAW: io.Reader | string | []byte | []string | <-chan string
	HTTPClient *http.Client
}

// TransportHandlers receive transport events.
type TransportHandlers struct {
	OnText  func(text string)
	OnDone  func()
	OnError func(err error)
}

// TransportHandle can abort an in-flight transport.
type TransportHandle struct {
	abort func()
}

// Abort cancels the transport.
func (h *TransportHandle) Abort() {
	if h != nil && h.abort != nil {
		h.abort()
	}
}

// OpenTransport starts I/O on a goroutine and returns an abort handle.
func OpenTransport(req TransportRequest, handlers TransportHandlers) *TransportHandle {
	var aborted atomic.Bool
	ctx, cancel := context.WithCancel(context.Background())

	go func() {
		defer cancel()
		var err error
		kind := req.Kind
		if kind == "" {
			kind = TransportHTTP
		}
		switch kind {
		case TransportRAW:
			err = runRaw(req, handlers, &aborted)
		case TransportSSE:
			err = runSSE(ctx, req, handlers, &aborted)
		default:
			err = runHTTP(ctx, req, handlers, &aborted)
		}
		if aborted.Load() {
			if handlers.OnError != nil {
				handlers.OnError(fmt.Errorf("aborted"))
			}
			return
		}
		if err != nil {
			if handlers.OnError != nil {
				handlers.OnError(err)
			}
			return
		}
		if handlers.OnDone != nil {
			handlers.OnDone()
		}
	}()

	return &TransportHandle{abort: func() {
		aborted.Store(true)
		cancel()
	}}
}

func emitText(h TransportHandlers, text string) {
	if text != "" && h.OnText != nil {
		h.OnText(text)
	}
}

func runRaw(req TransportRequest, h TransportHandlers, aborted *atomic.Bool) error {
	if req.Source == nil {
		return fmt.Errorf("raw transport requires source")
	}
	switch src := req.Source.(type) {
	case string:
		if aborted.Load() {
			return fmt.Errorf("aborted")
		}
		emitText(h, src)
		return nil
	case []byte:
		if aborted.Load() {
			return fmt.Errorf("aborted")
		}
		emitText(h, string(src))
		return nil
	case io.Reader:
		return readReaderUTF8(src, h, aborted)
	case []string:
		for _, piece := range src {
			if aborted.Load() {
				return fmt.Errorf("aborted")
			}
			emitText(h, piece)
		}
		return nil
	case <-chan string:
		for piece := range src {
			if aborted.Load() {
				return fmt.Errorf("aborted")
			}
			emitText(h, piece)
		}
		return nil
	default:
		return fmt.Errorf("raw source type %T not supported (want io.Reader, string, []byte, []string, or <-chan string)", src)
	}
}

func readReaderUTF8(r io.Reader, h TransportHandlers, aborted *atomic.Bool) error {
	buf := make([]byte, 8192)
	var carry []byte
	for {
		if aborted.Load() {
			return fmt.Errorf("aborted")
		}
		n, err := r.Read(buf)
		if n > 0 {
			carry = append(carry, buf[:n]...)
			for len(carry) > 0 {
				r, size := utf8.DecodeRune(carry)
				if r == utf8.RuneError && size == 1 {
					if len(carry) < 4 {
						break // wait for more bytes
					}
					emitText(h, string(utf8.RuneError))
					carry = carry[1:]
					continue
				}
				emitText(h, string(carry[:size]))
				carry = carry[size:]
			}
		}
		if err == io.EOF {
			if len(carry) > 0 {
				emitText(h, string(carry))
			}
			return nil
		}
		if err != nil {
			return err
		}
	}
}

func httpClient(req TransportRequest) *http.Client {
	if req.HTTPClient != nil {
		return req.HTTPClient
	}
	return http.DefaultClient
}

func runHTTP(ctx context.Context, req TransportRequest, h TransportHandlers, aborted *atomic.Bool) error {
	if req.URL == "" {
		return fmt.Errorf("transport url is required")
	}
	method := req.Method
	if method == "" {
		method = http.MethodGet
	}
	var body io.Reader
	if req.Body != "" {
		body = strings.NewReader(req.Body)
	}
	httpReq, err := http.NewRequestWithContext(ctx, method, req.URL, body)
	if err != nil {
		return err
	}
	for k, v := range req.Headers {
		httpReq.Header.Set(k, v)
	}
	resp, err := httpClient(req).Do(httpReq)
	if err != nil {
		if aborted.Load() || ctx.Err() != nil {
			return fmt.Errorf("aborted")
		}
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	return readReaderUTF8(resp.Body, h, aborted)
}

func runSSE(ctx context.Context, req TransportRequest, h TransportHandlers, aborted *atomic.Bool) error {
	headers := map[string]string{}
	for k, v := range req.Headers {
		headers[k] = v
	}
	if _, ok := headers["Accept"]; !ok {
		headers["Accept"] = "text/event-stream"
	}
	inner := TransportRequest{
		URL:        req.URL,
		Kind:       TransportHTTP,
		Method:     req.Method,
		Headers:    headers,
		Body:       req.Body,
		TimeoutMs:  req.TimeoutMs,
		HTTPClient: req.HTTPClient,
	}
	var mu sync.Mutex
	var buf strings.Builder
	wrapped := TransportHandlers{
		OnText: func(text string) {
			mu.Lock()
			defer mu.Unlock()
			buf.WriteString(text)
			flushSSEBlocks(&buf, h, req.SSEEvents)
		},
		OnDone:  func() {},
		OnError: h.OnError,
	}
	if err := runHTTP(ctx, inner, wrapped, aborted); err != nil {
		return err
	}
	mu.Lock()
	defer mu.Unlock()
	if rem := strings.TrimSpace(buf.String()); rem != "" {
		emitSSEData(h, ParseSSEBlock(buf.String(), req.SSEEvents))
	}
	return nil
}

func flushSSEBlocks(buf *strings.Builder, h TransportHandlers, allow map[string]struct{}) {
	s := buf.String()
	parts := splitOnBlankLines(s)
	if len(parts) <= 1 {
		return
	}
	buf.Reset()
	buf.WriteString(parts[len(parts)-1])
	for i := 0; i < len(parts)-1; i++ {
		emitSSEData(h, ParseSSEBlock(parts[i], allow))
	}
}

// splitOnBlankLines splits on \n\n or \r\n\r\n; last element is the incomplete tail.
func splitOnBlankLines(s string) []string {
	var parts []string
	start := 0
	i := 0
	for i < len(s) {
		if s[i] == '\n' && i+1 < len(s) && s[i+1] == '\n' {
			parts = append(parts, s[start:i+1])
			start = i + 2
			i += 2
			continue
		}
		if s[i] == '\n' && i >= 1 && s[i-1] == '\r' &&
			i+2 < len(s) && s[i+1] == '\r' && s[i+2] == '\n' {
			parts = append(parts, s[start:i+1])
			start = i + 3
			i += 3
			continue
		}
		i++
	}
	parts = append(parts, s[start:])
	return parts
}

func emitSSEData(h TransportHandlers, data string) {
	if data == "" {
		return
	}
	if !strings.HasSuffix(data, "\n") {
		data += "\n"
	}
	emitText(h, data)
}

// ParseSSEBlock joins multi-line data: fields with \n (Node SSE parser).
func ParseSSEBlock(block string, allow map[string]struct{}) string {
	if strings.TrimSpace(block) == "" {
		return ""
	}
	event := "message"
	var dataLines []string
	sc := bufio.NewScanner(strings.NewReader(block))
	for sc.Scan() {
		raw := sc.Text()
		if raw == "" || strings.HasPrefix(raw, ":") {
			continue
		}
		field, value := raw, ""
		if idx := strings.IndexByte(raw, ':'); idx >= 0 {
			field = raw[:idx]
			value = raw[idx+1:]
			if strings.HasPrefix(value, " ") {
				value = value[1:]
			}
		}
		switch field {
		case "event":
			event = value
		case "data":
			dataLines = append(dataLines, value)
		}
	}
	if allow != nil {
		if _, ok := allow[event]; !ok {
			return ""
		}
	}
	if len(dataLines) == 0 {
		return ""
	}
	return strings.Join(dataLines, "\n")
}

// ChunksOf is a RAW source helper (string pieces).
func ChunksOf(pieces ...string) []string {
	return pieces
}

// BytesReader wraps bytes as an io.Reader for RAW.
func BytesReader(b []byte) io.Reader {
	return bytes.NewReader(b)
}
